### PR TITLE
Improve the UI for selecting a user's location

### PR DESCRIPTION
**Pull request by esoergel**
@sravfeyn
The main functionality change here is that now you can deselect a user's location.
I've got a bunch of location changes coming through, mostly based on this [list](https://docs.google.com/a/dimagi.com/spreadsheets/d/1F-dZ-I31ItuQoClyalUiaS5oFu5WaZiZhtDFMBUOagk/edit#gid=0)

### CODE REVIEW

File: corehq/apps/commtrack/views.py [hunk 1]
```diff
@@ -168,42 +169,26 @@ def post(self, request, *args, **kwargs):
...
-
+            loc = SQLLocation.objects.get(location_id=id)
+        except SQLLocation.DoesNotExist:
+            return json_response(
+                {'message': 'no location with id %s found' % id},
+                status_code=404,
+            )
+        else:
+            return json_response(loc_to_payload(loc))
     else:
```
Comment (esoergel) [inline corehq/apps/commtrack/views.py]: I switched this to use SQLLocation.  Now it supports infix lookups too ("ass" will match "Massachusetts")

File: corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html [hunk 2]
```diff
@@ -7,8 +7,8 @@
 $(function() {
 
 $("#{{ id }}").select2({
-    placeholder: "Enter Name of Location",
-    minimumInputLength: 1,
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html]: this option was pretty annoying, and entirely unneeded

File: corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html [hunk 3]
```diff
@@ -35,15 +35,7 @@
     formatSelection: function(e) { return e.name; }
 });
 
-$('#clear_location').click(function() {
-    $("#{{ id }}").val('');
-    $("#commtrack_form").submit();
-})
-
```
Comment (esoergel) [inline corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html]: this stuff wasn't really being used.

File: corehq/apps/users/forms.py [hunk 4]
```diff
@@ -428,24 +417,10 @@ def __init__(self, *args, **kwargs):
...
-            user.set_location(loc)
-
-            # add the supply point case id to user data fields
-            # so that the phone can auto select
-            supply_point = SupplyPointCase.get_by_location(loc)
-            if supply_point:
-                user.user_data['commtrack-supply-point'] = supply_point._id
-
-            user.user_data['commcare_primary_case_sharing_id'] = \
-                LOCATION_SHARING_PREFIX + location_id
```
Comment (esoergel) [inline corehq/apps/users/forms.py]: This stuff is a duplication of what happens in `user.set_location`, plus `user` wasn't being saved after calling this anyways.

File: corehq/apps/users/models.py [hunk 5]
```diff
@@ -1334,6 +1334,13 @@ def set_location(self, location):
         Set the location, and all important user data, for
         the user.
         """
+        if not location:
+            # Remove associated info
+            self.user_data.pop('commtrack-supply-point', None)
+            self.user_data.pop('commcare_primary_case_sharing_id', None)
+            self.location_id = None
+            return
+
```
Comment (esoergel) [inline corehq/apps/users/models.py]: @twymer does this look right to you?  There are some `remove_location` type stuff in this same mixin, but they seemed like they were for a different thing.

Comment (sravfeyn) [inline corehq/apps/users/models.py]: Isn't popping some data in a function named `set_location` misleading?

**Comment by sravfeyn** [comment]
Looks good, will let @twymer respond and merge.

File: corehq/apps/users/models.py (hunk 5 above)
Comment (esoergel) [inline corehq/apps/users/models.py]: It only does that if you do `user.set_location(None)`, in which case it basically undoes whatever happens when you set a true location.

File: corehq/apps/commtrack/views.py (hunk 1 above)
Comment (czue) [inline corehq/apps/commtrack/views.py]: :unamused:

File: corehq/apps/commtrack/views.py [hunk 6]
```diff
@@ -168,42 +169,26 @@ def post(self, request, *args, **kwargs):
...
-
-        locs = sorted(
-            itertools.chain(*(get_locs(loc_type) for loc_type in loc_types)),
-            key=lambda e: e.name
-        )[:LIMIT]
-        return HttpResponse(json.dumps(map(loc_to_payload, locs)), 'text/json')
+        locs = SQLLocation.objects.filter(domain=domain)
+        if query:
+            locs = locs.filter(name__icontains=query)
+        return json_response(map(loc_to_payload, locs[:10]))
```
Comment (czue) [inline corehq/apps/commtrack/views.py]: dear god, this is so much better.

File: corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html (hunk 2 above)
Comment (czue) [inline corehq/apps/locations/templates/locations/manage/partials/autocomplete_select_widget.html]: :+1:

File: corehq/apps/users/forms.py (hunk 4 above)
Comment (czue) [inline corehq/apps/users/forms.py]: wtf

File: corehq/apps/users/models.py (hunk 5 above)
Comment (czue) [inline corehq/apps/users/models.py]: would suggest renaming this to `clear_location()` so it's more obvious and explict from an api perspective than `set_location(None)`

Comment (czue) [inline corehq/apps/users/models.py]: I think you might need to add a check that is something like:
```
if self.project.supports_multiple_locations_per_user:
    self.clear_locations()
```

though it also looks like there might be some problems with that function

Comment (czue) [inline corehq/apps/users/models.py]: http://manage.dimagi.com/default.asp?164671 for the vague thing

**Comment by czue** [comment]
@esoergel looks like a great change and I don't necessarily think anything of mine is blocking. I would like to get some clarity on how clearing should behave with multiple locations enabled though. I think what I proposed is "right" even if the function itself needs to be updated. i also want to rename the `add_location` and `remove_location` functions to make it more clear that they are only dealing with the delegate cases, but that can also be out of scope for this PR.

basically saying I'm ok with this being merged once you update the clear functionality to handle projects that allow multiple locations.

**Comment by esoergel** [comment]
@czue @twymer Do we actually support multiple locations per user or intend to?  If so, we could make this widget be a multiselect when that flag is checked.  If that's just going to stay a special case for a client, then maybe we should just disable this widget entirely for those domains, 'cause I don't think it really supports it well.

**Comment by esoergel** [comment]
Also, I see some comments in `LocationUserMixin` talking about a migration involving delegate cases, know what that's about?  AFAIK, here are the associations between a user and a location:

* user data `commtrack_supply_point` (id of the supply-point case, and only if location "tracks stock")
* user data `commcare_primary_case_sharing_id` (id of case-sharing group associated with location)
* location map case (case one-to-one with user which links one or more supply-point cases, also only if they track stock)

File: corehq/apps/users/models.py [hunk 7]
```diff
@@ -1366,6 +1366,16 @@ def set_location(self, location):
...
 
+    def unset_location(self):
+        """
+        Unset the location and remove all associated user data and cases
+        """
+        self.user_data.pop('commtrack-supply-point', None)
+        self.user_data.pop('commcare_primary_case_sharing_id', None)
+        self.location_id = None
+        self.clear_locations()
+        self.save()
```
Comment (esoergel) [inline corehq/apps/users/models.py]: @czue wrote:
> I think you might need to add a check that is something like:
>```
if self.project.supports_multiple_locations_per_user:
    self.clear_locations()
```
> though it also looks like there might be some problems with that function
I think I actually need to call `clear_locations` regardless of multiple locations - even with a single location, there is a location_map_case being created (perhaps needlessly?).

**Comment by czue** [comment]
we don't intend to support multiple locations but we can't break intrahealth (which uses it). the delegate case and location map case are the same hting

File: corehq/apps/users/models.py (hunk 7 above)
Comment (czue) [inline corehq/apps/users/models.py]: ah interesting. it might not be needless, but I'd have to dig around for a while to verify that. if it's being created then it should likely get deleted on clear so I think that should be safe.

**Comment by czue** [comment]
@esoergel I think this is good to go from my side. are you comfortable merging? one thing I wasn't sure about was whether we should QA the intrahealth case more when they clear a location in case this introduces problems. is it right that prior to this change there was no way for a user to clear their location?

**Comment by esoergel** [comment]
@czue that's correct, there was no way for a user to clear their location.  I tested out the multiple locations flag a bit, and it looks good to go in my opinion.  Here's how it works:

Each user has a primary location (same as a normal domain), and a list of locations.  After the flag gets set, any locations the user is assigned to here become the user's primary location, and are added on to the user's locations, unless the location doesn't track stock.  It doesn't look like it's possible (by design, I think) to assign a user to multiple locations which aren't supply-points.  This behavior is unchanged.

With this change, if the user "unsets" their location in this UI, it will remove the primary location from the user, as well as all of their other locations.  

That all sounds good to me, if a little confusing.  If you agree, I think we're g2m.

File: corehq/apps/users/models.py (hunk 7 above)
Comment (twymer) [inline corehq/apps/users/models.py]: This is legit. The delegate case is still how the supply point case lands on the phone for a user, sadly.

File: corehq/apps/commtrack/views.py (hunk 6 above)
Comment (twymer) [inline corehq/apps/commtrack/views.py]: :star: :+1: :fireworks: :green_heart:
